prow/plugins/blockade/blockade.go

```diff
@@ -0,0 +1,253 @@
+- repos:
+  - kubernetes/test-infra
+  blockregexps:
+  - "mungegithub/.*"
+  exceptionregexps:
+  - "mungegithub/DeprecationWarning.md"
+  explanation: "Don't work on mungegithub! Work on Prow!"
+*/
+
+package blockade
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName        = "blockade"
+	blockedPathsLabel = "do-not-merge/blocked-paths"
+)
+
+var blockedPathsBody = fmt.Sprintf("Adding label: `%s` because PR changes a protected file.", blockedPathsLabel)
+
+type githubClient interface {
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	CreateComment(org, repo string, number int, comment string) error
+}
+
+type pruneClient interface {
+	PruneComments(func(ic github.IssueComment) bool)
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest)
+}
+
+type blockCalc func([]github.PullRequestChange, []blockade) summary
+
+type client struct {
+	ghc    githubClient
+	log    *logrus.Entry
+	pruner pruneClient
+
+	blockCalc blockCalc
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	c := &client{
+		ghc:    pc.GitHubClient,
+		log:    pc.Logger,
+		pruner: pc.CommentPruner,
+
+		blockCalc: calculateBlocks,
+	}
+	return handle(c, pc.PluginConfig.Blockades, &pre)
+}
+
+// blockade is a compiled version of a plugins.Blockade config struct.
+type blockade struct {
+	blockRegexps, exceptionRegexps []*regexp.Regexp
+	explanation                    string
+}
+
+func (bd *blockade) isBlocker(file string) bool {
+	return matchesAny(file, bd.blockRegexps) && !matchesAny(file, bd.exceptionRegexps)
+}
+
+type summary map[string][]github.PullRequestChange
+
+func (s summary) String() string {
+	if len(s) == 0 {
+		return ""
+	}
+	var buf bytes.Buffer
+	fmt.Fprint(&buf, "#### Reasons for blocking this PR:\n")
+	for reason, files := range s {
+		fmt.Fprintf(&buf, "[%s]\n", reason)
+		for _, file := range files {
+			fmt.Fprintf(&buf, "- [%s](%s)\n\n", file.Filename, file.BlobURL)
+		}
+	}
+	return buf.String()
+}
+
+func handle(c *client, config []plugins.Blockade, pre *github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionSynchronize &&
+		pre.Action != github.PullRequestActionOpened &&
+		pre.Action != github.PullRequestActionReopened {
+		return nil
+	}
+
+	org := pre.Repo.Owner.Login
+	repo := pre.Repo.Name
+	labels, err := c.ghc.GetIssueLabels(org, repo, pre.Number)
+	if err != nil {
+		return err
+	}
+
+	labelPresent := hasBlockedLabel(labels)
+	blockades := compileApplicableBlockades(org, repo, c.log, config)
+	if len(blockades) == 0 && !labelPresent {
+		return nil
```

Comment:
`// if the label is missing, we assume that we removed any associated comments.`